### PR TITLE
Jobtime / Playtime Tracking

### PR DESCRIPTION
**Pull request by TheDarkElites**
## About the Pull Request

This will allow us to track player's time in jobs and then store that information on our server's database.
There is also now a playtime tab in the player panel to check a player's playtime for admins.
Roles can now be time-restricted.

P.S.: The reason the last pull request was closed was because it was a mess and I needed to start from scratch.

**Do Not Merge**

## Why It's Good For The Game

less LRP more HRP

## Current Time Requirements 
_(This is a baseline so tusurpta has something to work with when adjusting time requirements, and may not be the final requirements in game.)_

image
image

~~You can also find it in text form as follows:~~ Now Outdated, refer to picture above.

```
Site Director: 20 hrs Command
HOP: 2 hrs Command, 8hrs service, 8hrs bureacrat
Comms Officer: 10hrs Engineering, 10hrs Security, Comms tech 15hrs
Comms Tech: 5hrs Engineering
Internal Tribunal Department Officer: 10 hrs command, 15 hrs security, 2hrs bureacrat
Ethics Liaison: Command 3hrs, Bureacrat 8hrs
----------------------------------
Junior Engineer: Class D 30min
Engineer: 5hrs Junior Engineer
Senior Engineer: 10hrs Engineer, 12hrs Engineering
Containment Engineer: 10hrs Senior Engineer, 20hrs engineering
Chief Engineer: 10hrs Containment Engineer, 25hrs engineering
----------------------------------
EMT: Doctor 5hrs, Medical 10hrs
Surgeon: Doctor 8hrs, Medical 15hrs
Doctor: DClass 30min
Psychiatrist: Doctor 1hr
Chemist: DClass 30min
CMO: 2hrs EMT, 5hrs Surgeon, 10hrs Medical Doctor, 25hrs medical
----------------------------------
Logistics Officer: 5hr Logistics Specialist
Logistics Specialist: Class D 30min
----------------------------------
Archavist: 5hr Office Worker
----------------------------------
Junior Researcher: Class D 30min
Junior roboticist: Class D 30min
Researcher: 5hrs Junior Researcher
Roboticist: 5hrs Junior Roboticist
Senior Researcher: 10hrs Researcher, 12hrs Science
Senior Roboticist: 10hrs Roboticist, 10hrs Science
RD: 10hrs Senior Researcher, 25hrs Science
----------------------------------
Guard Commander: 5hr as each zone commander, 30hrs as security
Zone Commanders: 10hrs LCZ/HCZ/EZ(for their respective zone) Seargent/Senior Agent, 25hrs as LCZ/HCZ/EZ(for their respective zone)
Seargent/Senior Agent: 10hrs as LCZ/HCZ/EZ(for their respective zone) Guard/Agent, 12hrs as LCZ/HCZ/EZ(for their respective zone)
LCZ Guard: Class D 1hr
EZ Guard: LCZ 5hr
HCZ Guard: EZ 5hr
----------------------------------
AIC: 20hrs robot
```

## Changelog

:cl:
add: Jobtime Tracking and Timelocks for certain roles
server: new SQL table for role_time
code: goirep no longer has the COM flag, and is replaced with the new REP flag
admin: player panel now contains playtime tab
/:cl:

### CODE REVIEW

**Comment by TheDarkElites** [comment]
@Tsurupeta add do not merge label please

**Comment by Tsurupeta** [comment]
Nobody's merging drafts anyways

**Comment by TheDarkElites** [comment]
Okay this should be good for review.

**Comment by TheDarkElites** [comment]
@Tsurupeta When you review this, also look over the current set of time requirements and make any adjustments you see fit.

File: code/controllers/subsystems/jobs.dm [hunk 1]
```diff
@@ -10,6 +10,11 @@ var/const/SUP               =(1<<8)
 var/const/SPT               =(1<<9)
 var/const/EXP               =(1<<10)
 var/const/ROB               =(1<<11)
+var/const/LCZ               =(1<<12)
+var/const/ECZ               =(1<<13)
+var/const/HCZ               =(1<<14)
+var/const/BUR				=(1<<15)
+var/const/REP				=(1<<16)
```
Comment (Tsurupeta) [inline code/controllers/subsystems/jobs.dm]: replace all of this with defines.........

File: code/controllers/subsystems/jobtime.dm [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+	init_order = SS_INIT_JOBTIME
+	priority = SS_PRIORITY_JOBTIME
+
+	var/last_fired
+
+
+/datum/controller/subsystem/jobtime/Initialize()
+	if(!SSdbcore.IsConnected())
+		log_world("Jobtime tracking cannot function when DB is disconnected!")
+	last_fired = world.time
```
Comment (Tsurupeta) [inline code/controllers/subsystems/jobtime.dm]: And? Do you set canfire or what? You get the log and it continues working.

File: code/controllers/subsystems/jobtime.dm [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+	if(!SSdbcore.Connect())
+		return -1
+	for(var/client/L in GLOB.clients)
+		if(L.is_afk())
+			continue
+		L.update_jobtime_list(mins)
+
+//Client Procs
+
+/client/proc/update_jobtime_list(minutes)
```
Comment (Tsurupeta) [inline code/controllers/subsystems/jobtime.dm]: Does it all need to be defined on client?

File: code/controllers/subsystems/jobtime.dm [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+		if (!isnum(jvalue))
+			CRASH("invalid job value [jtype]:[jvalue]")
+		LAZYINITLIST(GLOB.jobtime_to_update)
+		GLOB.jobtime_to_update.Add(list(list(
+			"job" = jtype,
+			"ckey" = ckey,
+			"minutes" = jvalue)))
+	addtimer(CALLBACK(SSjobtime, /datum/controller/subsystem/jobtime/proc/update_jobtime_db),20,TIMER_OVERRIDE|TIMER_UNIQUE)
+
+/client/proc/get_jobtime_list_db()
```
Comment (Tsurupeta) [inline code/controllers/subsystems/jobtime.dm]: Along with this.

File: code/controllers/subsystems/jobs.dm (hunk 1 above)
Comment (TheDarkElites) [reply code/controllers/subsystems/jobs.dm:17]: That would mean throwing out the already existing department flags system (and everything that relies on it), unless we can do flags with defines somehow?

**Comment by TheDarkElites** [comment]
image
Since Tusurpta didn't actually sign off on it here, ill just put this as his review.

**Review by Tsurupeta — DISMISSED** [review]
Looks fine

**Comment by 77seven77seven** [comment]
wait wont this trap everyone into playing only class-D?

**Comment by TheDarkElites** [comment]
> wait wont this trap everyone into playing only class-D?

Not everyone, just new players, and only for 30 minutes. However, I'm going to go over the reqs list again and modify it to be cleaner. I'm probably only gonna keep the D-Class requirement for security.

**Comment by 77seven77seven** [comment]
> > wait wont this trap everyone into playing only class-D?
> 
> Not everyone, just new players, and only for 30 minutes. However, I'm going to go over the reqs list again and modify it to be cleaner. I'm probably only gonna keep the D-Class requirement for security.

ngl the D-class requirement for security is a good idea, this is going to make sec treat class-D better because they have to deal with sec doing the same shit to them

**Comment by TheDarkElites** [comment]
> > > wait wont this trap everyone into playing only class-D?
> > 
> > 
> > Not everyone, just new players, and only for 30 minutes. However, I'm going to go over the reqs list again and modify it to be cleaner. I'm probably only gonna keep the D-Class requirement for security.
> 
> ngl the D-class requirement for security is a good idea, this is going to make sec treat class-D better because they have to deal with sec doing the same shit to them

Yes that is the point.

**Comment by 77seven77seven** [comment]
> > > > wait wont this trap everyone into playing only class-D?
> > > 
> > > 
> > > Not everyone, just new players, and only for 30 minutes. However, I'm going to go over the reqs list again and modify it to be cleaner. I'm probably only gonna keep the D-Class requirement for security.
> > 
> > 
> > ngl the D-class requirement for security is a good idea, this is going to make sec treat class-D better because they have to deal with sec doing the same shit to them
> 
> Yes that is the point.

great minds think alike :trollface:

**Pull request merged**